subsys/net/lib/wifi_credentials/wifi_credentials.c

```diff
@@ -0,0 +1,444 @@
+
+int wifi_credentials_set_personal(
+	const char *ssid,
+	enum wifi_security_type security,
+	const uint8_t *bssid,
+	const char *psk,
+	const char *sae_password,
+	uint32_t flags
+)
+{
+	/* pack entry */
+	uint8_t buf[ENTRY_MAX_LEN] = {0};
+	struct wifi_credentials_header *header = (void *)buf;
+
+	header->security = security;
+	memcpy(header->ssid, ssid, WIFI_SSID_MAX_LEN);
+	header->flags = flags;
+
+	if (flags & WIFI_CREDENTIALS_FLAG_BSSID) {
+		if (bssid == NULL) {
+			return -EINVAL;
+		}
+		memcpy(header->bssid, bssid, WIFI_MAC_ADDR_LEN);
+	}
+
+	switch (security) {
+	case WIFI_SECURITY_TYPE_NONE:
+		break;
+	case WIFI_SECURITY_TYPE_PSK:
+	case WIFI_SECURITY_TYPE_PSK_SHA256:
+		if (psk == NULL) {
+			return -EINVAL;
+		}
+		struct wifi_credentials_psk *header_psk = (void *)buf;
+
+		memcpy(header_psk->psk, psk, WIFI_PSK_MAX_LEN);
+		break;
+	case WIFI_SECURITY_TYPE_SAE:
+		if (psk == NULL && sae_password == NULL) {
+			return -EINVAL;
+		}
+		struct wifi_credentials_sae *header_sae = (void *)buf;
+
+		if (psk != NULL) {
+			memcpy(header_sae->psk, psk, WIFI_PSK_MAX_LEN);
+		}
+		if (sae_password != NULL) {
+			memcpy(header_sae->sae_password, sae_password,
+			       CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH);
+		}
+		break;
+	default:
+		LOG_ERR("security type %d is unsupported", security);
+		return -ENOTSUP;
+	}
+
+	/* store entry */
+	return wifi_credentials_set_personal_raw(ssid, buf, wifi_credentials_expected_size(header));
+}
+
+int wifi_credentials_get_by_ssid_personal(
+	const char *ssid,
+	enum wifi_security_type *security,
+	uint8_t *bssid_buf,
+	char *psk_buf,
+	char *sae_password_buf,
+	uint32_t *flags
+)
+{
+	/* load entry */
+	uint8_t buf[ENTRY_MAX_LEN] = {0};
+	int ret = wifi_credentials_get_by_ssid_personal_raw(ssid, buf, ARRAY_SIZE(buf));
+
+	if (ret) {
+		return ret;
+	}
+
+	/* unpack entry*/
+	struct wifi_credentials_header *header = (void *)buf;
+
+	*security = header->security;
+	*flags = header->flags;
+	if (header->flags & WIFI_CREDENTIALS_FLAG_BSSID) {
+		if (bssid_buf == NULL) {
+			return -EINVAL;
+		}
+		memcpy(bssid_buf, header->bssid, WIFI_MAC_ADDR_LEN);
+	}
+
+	switch (header->security) {
+	case WIFI_SECURITY_TYPE_NONE:
+		break;
+	case WIFI_SECURITY_TYPE_PSK:
+	case WIFI_SECURITY_TYPE_PSK_SHA256:
+		if (psk_buf == NULL) {
+			return -EINVAL;
+		}
+		struct wifi_credentials_psk *header_psk = (void *)buf;
+
+		memcpy(psk_buf, header_psk->psk, WIFI_PSK_MAX_LEN);
+		break;
+	case WIFI_SECURITY_TYPE_SAE:
+		if (psk_buf == NULL && sae_password_buf == NULL) {
+			return -EINVAL;
+		}
+		struct wifi_credentials_sae *header_sae = (void *)buf;
+
+		if (psk_buf != NULL) {
+			memcpy(psk_buf, header_sae->psk, WIFI_PSK_MAX_LEN);
+		}
+		if (sae_password_buf != NULL) {
+			memcpy(sae_password_buf, header_sae->sae_password,
+			       CONFIG_WIFI_CREDENTIALS_SAE_PASSWORD_LENGTH);
+		}
+		break;
+	default:
+		LOG_ERR("data entry corrupted");
+		return -EFAULT;
+	}
+
+	return 0;
+}
+
+int wifi_credentials_delete_by_ssid(
```

Comment:
This and other functions are not thread safe. This could be problematic if we have, let's say, a BLE Provisioning Service and a connection manager accessing this data.

Reply:
Good point. I will try to make it thread-safe with a simple mutex.